builtin/plugins/dposv3/dpos.go

```diff
@@ -1349,15 +1355,53 @@ func ShiftDowntimeWindow(ctx contract.Context, currentHeight int64, candidates [
 	if state.Params.DowntimePeriod != 0 && (uint64(currentHeight)%state.Params.DowntimePeriod) == 0 {
 		ctx.Logger().Info("DPOS ShiftDowntimeWindow", "block", currentHeight)
 
+		downtimeSlashingEnabled := ctx.FeatureEnabled(loomchain.DPOSVersion3_4, false)
+
+		maxDowntimePercentage := defaultMaxDowntimePercentage
+		if state.Params.MaxDowntimePercentage != nil {
+			maxDowntimePercentage = state.Params.MaxDowntimePercentage.Value
+		}
+
+		maximumMissedBlocksBig := CalculateFraction(maxDowntimePercentage, loom.BigUInt{big.NewInt(int64(state.Params.DowntimePeriod))})
+		maximumMissedBlocks := maximumMissedBlocksBig.Uint64()
+
 		for _, candidate := range candidates {
-			statistic, err := GetStatistic(ctx, loom.UnmarshalAddressPB(candidate.Address))
+			candidateAddress := loom.UnmarshalAddressPB(candidate.Address)
+			statistic, err := GetStatistic(ctx, candidateAddress)
 			if err != nil {
 				if err == contract.ErrNotFound {
 					continue
 				}
 				return err
 			}
 
+			if downtimeSlashingEnabled {
+				slash := true;
+				downtime := getDowntimeRecord(ctx, statistic)
+				for i := uint64(0); i < 4; i++ {
+					if (maximumMissedBlocks >= downtime.Periods[i]) {
```

Comment:
Shouldn't we check only the last period? It seems that a validator is being slashed 4 times for a period that exceeds the maximumMissedBlock.

Reply:
No, the documents explain that if ANY of the last four periods include fewer missed blocks than the maximum, then there is no slashing. it's `slash := true` by default. It's a very lenient policy, very similar to what cosmos does.

Reply:
aha, my bad, this is where we unslash not slash